### DISCUSSION TITLE
Additional costs on multiple inferences in a row in the same session

### DISCUSSION THREAD

**Discussion by Manutea** [microsoft/onnxruntime#17590] [Other Q&A]
Hello,

When making multiple inferences in a row in the same session, some inferences incur additional costs.
How can I correct this ? I'm using Cuda 11.7 and OnnxRuntime 1.15.0

I did a few tests and you can see a sort of patern
![Figure_3](https://github.com/microsoft/onnxruntime/assets/17413106/0767db81-c636-4d99-a220-2b3984ac3ec0)
![Figure_2](https://github.com/microsoft/onnxruntime/assets/17413106/087120c0-f1ed-4650-a7e7-4369da505370)
![Figure_1](https://github.com/microsoft/onnxruntime/assets/17413106/649c5675-c5d6-469d-ab78-89f9d41ce7d4)

```cpp
void onnx_benchmark_GPU(std::string &modelPath, std::string &inputTensorName, std::string &outputTensorName, int deviceId, int batch)
{
  std::vector<float> image(batch * 3 * 224 * 224, 150);
  std::vector<int64_t> inputDims = {batch, 3, 224, 224};
  std::vector<int64_t> outputDims = {batch, 1000};
 
  Ort::Env env(ORT_LOGGING_LEVEL_WARNING, "InferenceGPU");
 
  Ort::SessionOptions sessionOptions;
  sessionOptions.EnableProfiling("gpu_profile_file");
  OrtStatus* status = OrtSessionOptionsAppendExecutionProvider_CUDA(sessionOptions, deviceId);

  Ort::Session session(env, modelPath.c_str(), sessionOptions);

  Ort::MemoryInfo infoCuda("Cuda", OrtAllocatorType::OrtArenaAllocator, deviceId, OrtMemTypeDefault);  
  Ort::Allocator cudaAllocator(session, infoCuda);
  
  int num_iterations = 4096;
  for (int i = 0; i < num_iterations; i++) 
  {
    Ort::IoBinding binding(session);
    auto input = cudaAllocator.GetAllocation(image.size() * sizeof(float));
    cudaMemcpy(input.get(), image.data(), sizeof(float) * image.size(), cudaMemcpyHostToDevice);

    // Create an OrtValue tensor backed by data on CUDA memory
    Ort::Value boundX = Ort::Value::CreateTensor(infoCuda, reinterpret_cast<float*>(input.get()), image.size(), inputDims.data(), inputDims.size());
    std::vector<float> outputData(std::accumulate(outputDims.begin(), outputDims.end(), 1, std::multiplies<int>()));
    auto output = cudaAllocator.GetAllocation(outputData.size() * sizeof(float));

    // Create an OrtValue tensor backed by data on CUDA memory
    Ort::Value boundY = Ort::Value::CreateTensor(infoCuda, reinterpret_cast<float*>(output.get()), outputData.size(), outputDims.data(), outputDims.size());
    
    binding.BindInput(inputTensorName.c_str(), boundX);
    binding.BindOutput(outputTensorName.c_str(), boundY);
    binding.SynchronizeInputs();
    session.Run(Ort::RunOptions(), binding);
    binding.SynchronizeOutputs();
    binding.ClearBoundInputs();
    binding.ClearBoundOutputs();
  }
}
```
I also use the tool /onnxruntime_perf_test :
./onnxruntime_perf_test -I -S 1 -e cuda -r 2048 -p profile.json -s /data/model/googlenet/dynamic_batch_googlenet_opt.onnx
![Screenshot from 2023-09-20 14-01-35](https://github.com/microsoft/onnxruntime/assets/17413106/d08ddd40-4727-4693-90e8-c7d94f03b6e1)

And for the task Id ~1000, 2000, 4000, 8000, 16000 the bug doubles latency.
Maybe it's Cuda  or something reallocate from time to time, and in larger and larger quantities to limit the extra cost.

![Figure_1](https://github.com/microsoft/onnxruntime/assets/17413106/84b3be3c-1fd6-4948-9c1b-b3ea0017ea4d)

Thanks in advance

**Comment by gh-yewang** [comment]
could be related with https://github.com/microsoft/onnxruntime/pull/14023?
please consider raising an issue

**Reply by Manutea** [reply]
Thank you, an issue is open [#17720](https://github.com/microsoft/onnxruntime/issues/17720)
